pkg/asset/machines/openstack/machinesets.go

```diff
@@ -42,12 +42,29 @@ func MachineSets(clusterID string, config *types.InstallConfig, pool *types.Mach
 	numOfAZs := int32(len(mpool.Zones))
 	var machinesets []*clusterapi.MachineSet
 
-	for idx, az := range mpool.Zones {
+	var subnets, zones []string
+	if len(mpool.FailureDomainNames) > 0 {
+		for _, name := range mpool.FailureDomainNames {
+			if len(platform.FailureDomains) > 0 {
+				for _, domain := range platform.FailureDomains {
+					if domain.Name == name {
+						zones = append(zones, domain.ComputeZone)
+						volumeAZs = append(volumeAZs, domain.StorageZone)
+						subnets = append(subnets, domain.Subnet)
+						break
+					}
+				}
+			}
+		}
+	} else {
+		zones = mpool.Zones
+	}
+	for idx, az := range zones {
 		replicas := int32(total / numOfAZs)
 		if int32(idx) < total%numOfAZs {
 			replicas++
 		}
-		provider, err := generateProvider(clusterID, platform, mpool, osImage, az, role, userDataSecret, trunkSupport, volumeAZs[idx%len(volumeAZs)])
+		provider, err := generateProvider(clusterID, platform, mpool, osImage, az, role, userDataSecret, trunkSupport, volumeAZs[idx%len(volumeAZs)], subnets[idx%len(subnets)])
```

Comment:
this somehow creates too many machinesets. If I have 3 FailureDomains for the workers, it will create 9 workers, I need to figure that out.